aptos-move/framework/aptos-framework/sources/transaction_fee.move

```diff
@@ -1,15 +1,104 @@
+        burn_percentage: u8,
+    }
+
+    /// Initializes the resource storing information about gas fees collection and
+    /// distribution. Should be called by on-chain governance.
+    public fun initialize_fee_collection_and_distribution(aptos_framework: &signer, burn_percentage: u8) {
+        // Sanity checks.
+        system_addresses::assert_aptos_framework(aptos_framework);
+        assert!(
+            !exists<CollectedFeesPerBlock>(@aptos_framework),
+            error::already_exists(EALREADY_COLLECTING_FEES)
+        );
+        assert!(burn_percentage <= 100, error::out_of_range(EINVALID_BURN_PERCENTAGE));
+
+        // Make sure stakng module is aware of transaction fees collection.
+        stake::initialize_validator_fees(aptos_framework);
+
+        // Initially, no fees are collected and the block proposer is not set.
+        let zero = coin::initialize_aggregatable_coin(aptos_framework);
+        let info = CollectedFeesPerBlock {
+            amount: zero,
+            proposer: option::none(),
+            burn_percentage,
+        };
+        move_to(aptos_framework, info);
+    }
+
+    /// Registers the proposer of the block for gas fees collection. This function
+    /// can only be called at the beginning of the block.
+    public(friend) fun register_proposer_for_fee_collection(proposer_addr: address) acquires CollectedFeesPerBlock {
+        let collected_fees = borrow_global_mut<CollectedFeesPerBlock>(@aptos_framework);
+        let _ = option::swap_or_fill(&mut collected_fees.proposer, proposer_addr);
+    }
+
+    /// Burns a specified fraction of the coin.
+    fun burn_coin_fraction(coin: &mut Coin<AptosCoin>, burn_percentage: u8) acquires AptosCoinCapabilities {
+        assert!(burn_percentage <= 100, error::out_of_range(EINVALID_BURN_PERCENTAGE));
+
+        let collected_amount = coin::value(coin);
+        let amount_to_burn = (burn_percentage as u64) * collected_amount / 100;
+        let coin_to_burn = coin::extract(coin, amount_to_burn);
+        burn_collected_fee(coin_to_burn);
+    }
+
+    /// Calculates the fee which should be distributed to the block proposer at the
+    /// end of an epoch, and records it in the system. This function can only be called
+    /// at the beginning of the block or during reconfiguration.
+    public(friend) fun assign_or_burn_collected_fee() acquires AptosCoinCapabilities, CollectedFeesPerBlock {
+        let collected_fees = borrow_global_mut<CollectedFeesPerBlock>(@aptos_framework);
+
+        // If there are no collected fees, do nothing.
+        if (coin::is_aggregatable_coin_zero(&collected_fees.amount)) {
+            return
+        };
+
+        // Otherwise get the collected fee, and check if it can distributed later.
+        let coin = coin::drain_aggregatable_coin(&mut collected_fees.amount);
+        if (option::is_some(&collected_fees.proposer)) {
+            let proposer_addr = *option::borrow(&collected_fees.proposer);
```

Comment:
Shouldn't this remove the proposer from collected_fees now that the fees have been sent to the stake module to track for this proposer?

Reply:
I separated this into a sand-alone function `register_proposer_for_fee_collection` which overwrites the proposer anyway. You think explicitly setting to `option::none()` would be better?

Reply:
Yeah explicitly setting it to option::none() would be clearer that there's no lingering tie to the previous proposer who received the fee distribution